nvflare/app_common/ccwf/swarm_client_ctl.py

```diff
@@ -0,0 +1,620 @@
+            )
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        if result_round > self.for_round:
+            # this should never happen!
+            # otherwise it means that the client is sending me result for a round that I couldn't possibly schedule!
+            self.log_error(
+                fl_ctx,
+                f"logic error: received result from {client_name} for round {result_round}, "
+                f"which is > gatherer's current round {self.for_round}",
+            )
+            self.executor.update_status(action="gather", error=ReturnCode.EXECUTION_EXCEPTION)
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        if result_round < self.for_round:
+            # this is a late result for a round that I scheduled in the past.
+            # Note: we still accept it!
+            self.log_warning(
+                fl_ctx,
+                f"received late result from {client_name} for round {result_round}, "
+                f"which is < gatherer's current round {self.for_round}",
+            )
+
+        if result_round == self.for_round:
+            # this is the result that I'm waiting for.
+            now = time.time()
+            ts.reply_time = now
+            if not self.min_resps_received_time:
+                # see how many responses I have received
+                num_resps_received = 0
+                for _, ts in self.trainer_statuses.items():
+                    if ts.reply_time:
+                        num_resps_received += 1
+                if num_resps_received >= self.min_responses_required:
+                    self.min_resps_received_time = now
+
+        rc = result.get_return_code(ReturnCode.OK)
+        if rc != ReturnCode.OK:
+            self.log_error(fl_ctx, f"Bad result from {client_name} for round {result_round}: {rc}.")
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        fl_ctx.set_prop(AppConstants.CURRENT_ROUND, self.for_round, private=True, sticky=True)
+        fl_ctx.set_prop(AppConstants.TRAINING_RESULT, result, private=True, sticky=False)
+        self.fire_event(AppEventType.BEFORE_CONTRIBUTION_ACCEPT, fl_ctx)
+
+        accepted = self.aggregator.accept(result, fl_ctx)
+        accepted_msg = "ACCEPTED" if accepted else "REJECTED"
+        self.log_info(
+            fl_ctx, f"Contribution from {client_name} {accepted_msg} by the aggregator at round {result_round}."
+        )
+
+        fl_ctx.set_prop(AppConstants.AGGREGATION_ACCEPTED, accepted, private=True, sticky=False)
+        self.fire_event(AppEventType.AFTER_CONTRIBUTION_ACCEPT, fl_ctx)
+        return make_reply(ReturnCode.OK)
+
+    def aggregate(self):
+        fl_ctx = self.fl_ctx
+        self.log_info(fl_ctx, f"Start aggregation for round {self.for_round}")
+        self.fire_event(AppEventType.BEFORE_AGGREGATION, fl_ctx)
+        aggr_result = self.aggregator.aggregate(fl_ctx)
```

Comment:
We have to, that's how other widgets (e.g. model selector, shareable generator) are designed to work.
BUT this doesn't mean that the aggr result CONTENT has to be in memory. It could very well be just a file name. Of course we may need to develop some file-based components. But these are out of the scope of this PR.